src/daemon/custom_image_host.cpp

```diff
@@ -91,56 +97,77 @@ auto map_aliases_to_vm_info_for(const std::vector<mp::VMImageInfo>& images)
     std::unordered_map<std::string, const mp::VMImageInfo*> map;
     for (const auto& image : images)
     {
-        map[image.id.toStdString()] = &image;
-        for (const auto& alias : image.aliases)
+        // non default entry filtering after parallel writing
+        if (!is_default_constructed(image))
         {
-            map[alias.toStdString()] = &image;
+            map[image.id.toStdString()] = &image;
+            for (const auto& alias : image.aliases)
+            {
+                map[alias.toStdString()] = &image;
+            }
         }
     }
 
     return map;
 }
 
-auto full_image_info_for(const QMap<QString, CustomImageInfo>& custom_image_info, mp::URLDownloader* url_downloader)
+auto full_image_info_for(const QMap<QString, CustomImageInfo>& custom_image_info, mp::URLDownloader* url_downloader,
+                         bool is_force_update_from_network = false)
 {
-    std::vector<mp::VMImageInfo> default_images;
+    std::vector<mp::VMImageInfo> default_images(custom_image_info.size());
+
+    auto fetch_one_image_info_and_write_to_index =
+        [&default_images,
+         is_force_update_from_network](int index, mp::URLDownloader* url_downloader,
+                                       const std::pair<QString, CustomImageInfo>& image_info_pair) -> void {
+        const QString& image_file_name = image_info_pair.first;
+        const CustomImageInfo& custom_image_info = image_info_pair.second;
+        const QString image_url{custom_image_info.url_prefix + image_info_pair.first};
+        const QString hash_url{custom_image_info.url_prefix + QStringLiteral("SHA256SUMS")};
+
+        const auto base_image_info =
+            base_image_info_for(url_downloader, image_url, hash_url, image_file_name, is_force_update_from_network);
+
+        default_images[index] = mp::VMImageInfo{custom_image_info.aliases,
+                                                custom_image_info.os,
+                                                custom_image_info.release,
+                                                custom_image_info.release_string,
+                                                true,                 // supported
+                                                image_url,            // image_location
+                                                base_image_info.hash, // id
+                                                "",
+                                                base_image_info.last_modified, // version
+                                                0,
+                                                true};
+    };
 
-    for (const auto& image_info : custom_image_info.toStdMap())
+    std::vector<std::future<void>> empty_futures;
+    empty_futures.reserve(custom_image_info.size());
```

Comment:
That one is slightly different in terms of the behavior. `std::vector` constructor with size allocated `custom_image_info.size()` spaces and **defaultly constructed all the objects**, whereas `reserve` only allocate the raw spaces. This makes a difference to the later code `empty_futures.emplace_back(...);` if we go with `std::vector<std::future<void>> empty_futures(custom_image_info.size()));`, then the code will need to be changed to `empty_futures[index] = ...;` because `empty_futures` is already filled in with default objects, we only can overwrite it. I personally prefer the `reserve` here because the `std::future` object is lazily constructed.

Reply:
Ah, ok, that makes sense.  Thanks for the explanation and I'm good with this.